learn/introduction-to-scryer.dj

```diff
@@ -0,0 +1,319 @@
+# Introduction to Scryer Prolog
+
+This tutorial  is intended to help you get started with [Scryer Prolog](https://www.scryer.pl/), 
+and perhaps save you some time as you learn Prolog.  
+
+It is not intended to teach you Prolog concepts.  The [Power of Prolog](https://www.metalevel.at/prolog) is a great resource for that.    
+
+## Choose your Operating System
+
+As a Prolog learner, you should consider Linux or Mac OSX;  Windows is not as widely used in the Scyrer community.  Windows users 
+ are likely to run into problems with Scryer Linux or Mac users do not exerience.    
+
+## Avoid or Unlearn Legacy Predicates 
+
+Many tutorials or lessons will teach facets of Prolog that are best avoided, at least while you are at the early stages of learning the language.  Prolog has improved over its lifetime, but many recorded lectures or tutorials ignore or predate those advancements.
+
+As a beginner, these legacy Prolog predicates are sharp corners you may wish to avoid.  
+
+```prolog
+- (!) cut.  
+- (->) if.  
+(is).  
+(\+).
+include/3.
+exclude/3.
+```
+
+For more information see [Prolog Patterns:  How Not to Do It](https://www.youtube.com/watch?v=6G-3DqyJ_l8).
+
+
+## Install Scryer Prolog
+
+There are prebuilt downloads for Debian, macOS, and Windows on [www.scryer.pl](https://www.scryer.pl/).  This is the fastest way to get started.
+
+You may find you wish to build Scryer from source to incorporate the latest features and bug fixes.  It isn't terribly difficult.  Some tips:
+
+* keep your Rust current.
+* build Scryer for release, not debug.   The debug builds take a long time to load and run. 
+
+## Hello World
+
+Run a basic Scryer program.
+
+It is useful to know how to do something when a Prolog script loads.
+You may have tests you wish to run, and it can be tedious to run them
+in the Scryer toplevel every time your code changes.
+
+Copy this into hello.pl:
+
+```prolog
+:- use_module(library(format)).
+:- initialization(main).
+
+main :-
+    format("Hello, world!~n", []).
```

Comment:
This is a side-effect that cannot be tested easily. A much more testable solution is to describe the string with a DCG, and use `phrase_to_stream/2` for the output.

Reply:
My point was  one of the first things that can save time is to run a goal automatically when scryer starts.  It took me awhile to figure that out, and I often need it, if nothing else than to run through a bunch of tests.  

Are you suggesting it is bad to explain the use of initialization directive, or bad because of the use of format?

Reply:
In my opinion, the `initialization/1` directive is also not a good idea, since it prevents us from testing other predicates without running the given goal first. Personally, I would recommend to use the **`-g Goal`** command line switch, and a good name for such a predicate is&nbsp;**`run/0`**. So we would have:

<pre>
$ scryer-prolog <b>-g run</b> example.pl
</pre>

Or:

<pre>
$ scryer-prolog <b>-g run,halt</b> example.pl
</pre>

Or, with #3147 applied:

<pre>
$ scryer-prolog <b>-g run -t halt</b> example.pl
</pre>